### PR TITLE
Add iOSNativeLoginTemplate.

### PR DESCRIPTION
**Pull request by brandonpage**
Cloned `iOSNativeSwiftTemplate` with Native Login Added.  Native Login screen is a simplified version of my demo because the animations add a lot of LOC making it a less clear example.

### CODE REVIEW

File: iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginView.swift [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+//  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+//  FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+//  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+//  WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import SwiftUI
```
Comment (brandonpage) [inline iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginView.swift:28]: Native Login Example UI.

File: iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginViewFactory.swift [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+//  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+//  WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import Foundation
+import SwiftUI
+
+@objc class NativeLoginViewFactory: NSObject {
```
Comment (brandonpage) [inline iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginViewFactory.swift:31]: Example of how to get a ViewController from SwiftUI.

File: iOSNativeLoginTemplate/iOSNativeLoginTemplate/SceneDelegate.swift [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        // This Client Id and Redirect Url is specific and exclusive to Native Login.
+        let clientId = "3MVG9CEn_O3jvv0wTqRT0Le6tmzX.EQ9ZvtHL1TG3gHFV.4IvKZyXw5SgdiVPi61mXrpu40mCOhKevEfYNMOm"
+        let redirectUri = "https://msdk-enhanced-dev-ed.my.site.com/services/oauth2/echo"
+        let loginUrl = "https://msdk-enhanced-dev-ed.my.site.com/headless"
+        // Used to create a View Controller from SwiftUI.
+        let nativeLoginViewConroller = NativeLoginViewFactory.create()
+
+        // This line tells the SDK the app intends to use Native Login.
+        SalesforceManager.shared.useNativeLogin(withConsumerKey: clientId, callbackUrl: redirectUri, communityUrl: loginUrl,
+                                                nativeLoginViewController: nativeLoginViewConroller, scene:scene)
```
Comment (brandonpage) [inline iOSNativeLoginTemplate/iOSNativeLoginTemplate/SceneDelegate.swift:49]: Native Login Setup.

File: iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginViewFactory.swift [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    @objc static func create() -> UIViewController {
+        let view = UIHostingController(rootView: NativeLoginView())
+        view.viewDidLayoutSubviews()
```
Comment (bbirman) [inline iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginViewFactory.swift:35]: I thought this was usually called by the system, why is this needed?

File: iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginView.swift [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+import SalesforceSDKCore
+
+struct NativeLoginView: View {
+    @Environment(\.colorScheme) var colorScheme
+    
+    @State private var username = ""
+    @State private var password = ""
+    @State private var errorMessage = ""
+    @State private var isAuthenticating = false
+    var foreverAnimation: Animation {
```
Comment (bbirman) [inline iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginView.swift:38]: is this used?

File: iOSNativeLoginTemplate/iOSNativeLoginTemplate/SceneDelegate.swift [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    func sceneWillResignActive(_ scene: UIScene) {
+        // Called when the scene will move from an active state to an inactive state.
+        // This may occur due to temporary interruptions (ex. an incoming phone call).
+    }
+
+    func sceneWillEnterForeground(_ scene: UIScene) {
+        // Called as the scene transitions from the background to the foreground.
+        // Use this method to undo the changes made on entering the background.
+        self.initializeAppViewState()
+        AuthHelper.loginIfRequired {
```
Comment (bbirman) [inline iOSNativeLoginTemplate/iOSNativeLoginTemplate/SceneDelegate.swift:79]: how does this work with the native login flow?

File: iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginView.swift (hunk 5 above)
Comment (brandonpage) [reply iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginView.swift:38]: Nope, good catch.

File: iOSNativeLoginTemplate/iOSNativeLoginTemplate/SceneDelegate.swift (hunk 6 above)
Comment (brandonpage) [reply iOSNativeLoginTemplate/iOSNativeLoginTemplate/SceneDelegate.swift:79]: Yes, like webview login _it is required_.  Native Login is launched through the same `loginWithCompletion` codepath as the webview.

File: iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginViewFactory.swift (hunk 4 above)
Comment (brandonpage) [reply iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginViewFactory.swift:35]: I thought so but apparently not.  👍

**Review by wmathurin — CHANGES_REQUESTED** [review]
You should also add it to [templates.json](https://github.com/forcedotcom/SalesforceMobileSDK-Templates/blob/dev/templates.json) - so that `forceios listtemplates` can discover it.

File: iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginView.swift [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+struct NativeLoginView: View {
+    @Environment(\.colorScheme) var colorScheme
+    
+    @State private var username = ""
+    @State private var password = ""
+    @State private var errorMessage = ""
+    @State private var isAuthenticating = false
+    
+    init() {
```
Comment (bbirman) [inline iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginView.swift:39]: not sure if this is intentional for something else, you don't need to define init if it's empty

File: iOSNativeLoginTemplate/iOSNativeLoginTemplate/SceneDelegate.swift [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    var window: UIWindow?
+    
+    func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
+        
+        guard let windowScene = (scene as? UIWindowScene) else { return }
+        self.window = UIWindow(frame: windowScene.coordinateSpace.bounds)
+        self.window?.windowScene = windowScene
+        
+        // This Client Id and Redirect Url is specific and exclusive to Native Login.
+        let clientId = "3MVG9CEn_O3jvv0wTqRT0Le6tmzX.EQ9ZvtHL1TG3gHFV.4IvKZyXw5SgdiVPi61mXrpu40mCOhKevEfYNMOm"
```
Comment (wmathurin) [inline iOSNativeLoginTemplate/iOSNativeLoginTemplate/SceneDelegate.swift:41]: Will customer be able to login to their org or will they need to first edit those lines (after doing the necessary setup in their org)?

**Review by bbirman — APPROVED** [review]

File: iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginViewFactory.swift (hunk 2 above)
Comment (bbirman) [reply iOSNativeLoginTemplate/iOSNativeLoginTemplate/NativeLoginViewFactory.swift:31]: what's objc for?

File: iOSNativeLoginTemplate/iOSNativeLoginTemplate/SceneDelegate.swift (hunk 8 above)
Comment (brandonpage) [reply iOSNativeLoginTemplate/iOSNativeLoginTemplate/SceneDelegate.swift:41]: @wmathurin They will need to define their own so I will definitely update the comment here to state that.  Should I remove our values to make it more obvious?  (I could even Assert the are not empty).  If I leave them it makes testing easier for us but not that big of a deal.  CI can work around this too.

Comment (brandonpage) [reply iOSNativeLoginTemplate/iOSNativeLoginTemplate/SceneDelegate.swift:41]: Could be something like this:
```
        ///
        /// Fill in the values below from the connected app that was created for Native Login and
        /// the url of your Experience Cloud community.
        ///
        let clientId = "your-client-id"
        let redirectUri = "your-redirect-uri"
        let loginUrl = "your-community-url"
        
        assert(clientId != "your-client-id", "Please add your Native Login client id.")
        assert(redirectUri != "your-redirect-uri", "Please add your Native Login redirect uri.")
        assert(loginUrl != "your-community-url", "Please add your Native Login community url.")
```

We can also update it with a link to the documentation when it goes live.

Comment (wmathurin) [reply iOSNativeLoginTemplate/iOSNativeLoginTemplate/SceneDelegate.swift:41]: We might need to amend the next steps instructions that forceios prints out.
Right now it only mentions the bootconfig file:
https://github.com/forcedotcom/SalesforceMobileSDK-Package/blob/dev/shared/createHelper.js#L237

**Review by wmathurin — APPROVED** [review]
